scenario/state.py

```diff
@@ -828,20 +833,23 @@ class Container(_max_posargs(1)):
         }
     """
 
-    exec_mock: _ExecMock = dataclasses.field(default_factory=dict)
+    execs: Iterable[Exec] = frozenset()
     """Simulate executing commands in the container.
 
-    Specify each command the charm might run in the container and a :class:`ExecOutput`
+    Specify each command the charm might run in the container and an :class:`Exec`
     containing its return code and any stdout/stderr.
 
     For example::
 
         container = scenario.Container(
             name='foo',
-            exec_mock={
-                ('whoami', ): scenario.ExecOutput(return_code=0, stdout='ubuntu')
-                ('dig', '+short', 'canonical.com'):
-                    scenario.ExecOutput(return_code=0, stdout='185.125.190.20\\n185.125.190.21')
+            execs={
+                scenario.Exec(('whoami', ), return_code=0, stdout='ubuntu'),
```

Comment:
The 1-element tuples are always awkward, and I think just matching on a single string is going to be quite common. How about in all our examples and tests we use lists, like `['whoami']`? I think it's more common in Python to use lists for "list of the same type of object" and tuples for "record/struct of different types of fields" (apart from their mutability differences). Thoughts?

Reply:
> The 1-element tuples are always awkward,

:100:

> I think it's more common in Python to use lists for "list of the same type of object" and tuples for "record/struct of different types of fields" (apart from their mutability differences).

I think there's definitely a chunk of Python developers that do this (and the docs encourage it), but I think there's a reasonable number that don't approach them this way too. To be fair, tuples are considerably faster and a little more memory efficient too, as well as the mutability differences.

> and I think just matching on a single string is going to be quite common. How about in all our examples and tests we use lists, like `['whoami']`?

Sure, I'm fine with that.

Reply:
I remember now why I didn't have this previously: the mutability does matter. The `Exec` objects get put into a frozenset in the `Container` object, so they need to be immutable (well, hashable).

However, I'm still fine with having the examples use lists, and I think people will want to do that anyway. So I've added post-init 'freezing' of the provided sequence into a tuple. The downside is that you pass in a list and then do `Exec.command_prefix` and you get back a tuple, which is unexpected. However, we're already doing this elsewhere to avoid having to write `frozenset({...})` all over the place, so I think it's fine.